storage/postgres/log_storage.go

```diff
@@ -0,0 +1,1011 @@
+	var ret []*trillian.LogLeaf
+	for rows.Next() {
+		leaf := &trillian.LogLeaf{}
+		// We might be using a LEFT JOIN in our statement, so leaves which are
+		// queued but not yet integrated will have a NULL IntegrateTimestamp
+		// when there's no corresponding entry in SequencedLeafData, even though
+		// the table definition forbids that, so we use a nullable type here and
+		// check its validity below.
+		var integrateTS sql.NullInt64
+		var queueTS int64
+
+		if err := rows.Scan(&leaf.MerkleLeafHash, &leaf.LeafIdentityHash, &leaf.LeafValue, &leaf.LeafIndex, &leaf.ExtraData, &queueTS, &integrateTS); err != nil {
+			glog.Warningf("LogID: %d Scan() %s = %s", t.treeID, desc, err)
+			return nil, err
+		}
+		var err error
+		leaf.QueueTimestamp, err = ptypes.TimestampProto(time.Unix(0, queueTS))
+		if err != nil {
+			return nil, fmt.Errorf("got invalid queue timestamp: %v", err)
+		}
+		if integrateTS.Valid {
+			leaf.IntegrateTimestamp, err = ptypes.TimestampProto(time.Unix(0, integrateTS.Int64))
+			if err != nil {
+				return nil, fmt.Errorf("got invalid integrate timestamp: %v", err)
+			}
+		}
+
+		if got, want := len(leaf.MerkleLeafHash), t.hashSizeBytes; got != want {
+			return nil, fmt.Errorf("LogID: %d Scanned leaf %s does not have hash length %d, got %d", t.treeID, desc, want, got)
+		}
+
+		ret = append(ret, leaf)
+	}
+
+	return ret, nil
+}
+
+func (t *readOnlyLogTX) GetUnsequencedCounts(ctx context.Context) (storage.CountByLogID, error) {
+	stx, err := t.tx.PrepareContext(ctx, selectUnsequencedLeafCountSQL)
+	if err != nil {
+		glog.Warningf("Failed to prep unsequenced leaf count statement: %v", err)
+		return nil, err
+	}
+	defer stx.Close()
+
+	rows, err := stx.QueryContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	ret := make(map[int64]int64)
+	for rows.Next() {
+		var logID, count int64
+		if err := rows.Scan(&logID, &count); err != nil {
+			return nil, fmt.Errorf("failed to scan row from unsequenced counts: %v", err)
+		}
+		ret[logID] = count
+	}
+	return ret, nil
+}
+
+// leafAndPosition records original position before sort.
+type leafAndPosition struct {
+	leaf *trillian.LogLeaf
+	idx  int
+}
+
+// byLeafIdentityHashWithPosition allows sorting (as above), but where we need
+// to remember the original position
+type byLeafIdentityHashWithPosition []leafAndPosition
+
+func (l byLeafIdentityHashWithPosition) Len() int {
+	return len(l)
+}
+func (l byLeafIdentityHashWithPosition) Swap(i, j int) {
+	l[i], l[j] = l[j], l[i]
+}
+func (l byLeafIdentityHashWithPosition) Less(i, j int) bool {
+	return bytes.Compare(l[i].leaf.LeafIdentityHash, l[j].leaf.LeafIdentityHash) == -1
+}
+
+func isDuplicateErr(err error) bool {
+	switch err := err.(type) {
+	case *pq.Error:
+		glog.Infof("checking %v %v %v", err.Code, errNumDuplicate, err.Code == errNumDuplicate)
+		return err.Code == errNumDuplicate
+	default:
+		return false
+	}
+}
+
+// Copyright 2016 Google Inc. All Rights Reserved.
```

Comment:
Spurious file header here?

Reply:
I can make it go away; I'm sure it was some furious gittin I've been doing